code/parse/sexp.cpp

```diff
@@ -23094,29 +23094,32 @@ int eval_sexp(int cur_node, int referenced_node)
 	Assert( (type == SEXP_LIST) || (type == SEXP_ATOM) );
 
 	// trap known true and known false sexpressions.  We don't trap on SEXP_NAN sexpressions since
-	// they may yet evaluate to true or false.
+	// they may yet evaluate to true or false. If the sexp depends on the special argument,
+	// we can't 'know' its value so we skip this behaviour.
 
 	// we want to log event values for KNOWN_X or FOREVER_X before returning
-	if (Log_event && ((Sexp_nodes[cur_node].value == SEXP_KNOWN_TRUE) || (Sexp_nodes[cur_node].value == SEXP_KNOWN_FALSE) || (Sexp_nodes[cur_node].value == SEXP_NAN_FOREVER))) {
-		// if this is a node that has been assigned the value by short-circuiting,
-		// it might not be the operator that returned the value
-		int op_index = get_operator_index(cur_node);
-		if (op_index < 0)
-			op_index = get_operator_index(CAR(cur_node));
+	if (!(Sexp_nodes[cur_node].flags & SNF_SPECIAL_ARG_IN_NODE)) {
```

Comment:
Why are you only checking this one node instead of using `special_argument_appears_in_sexp_tree()`, as I thought you were planning based on the Discord discussion? Surely, if short-circuiting `( is-destroyed-delay "<argument>" 0 )` is bad, then short-circuiting `( or ( is-destroyed-delay "<argument>" 0 ) ( is-departed-delay "<argument>" 0 ) )` is also bad.

Reply:
Ack, you're right. I think I meant for SNF_SPECIAL_ARG_IN_TREE... but at this stage it'd be better just to go back to using special_argument_appears_in_sexp_tree().